src/mongo_odm/nvp.hpp

```diff
@@ -21,67 +21,686 @@
-        ar(cereal::make_nvp(nvp.name, this->*(nvp.t)));                                         \
-        mongo_odm_serialize_recur<Archive, I + 1, N>(ar);                                       \
-    }                                                                                           \
-                                                                                                \
-    template <class Archive, size_t I, size_t N>                                                \
-    typename std::enable_if<(I == N), void>::type mongo_odm_serialize_recur(Archive&) {         \
-        ;                                                                                       \
-    }
-
-// Register members and create serialize() function
-#define MONGO_ODM_MAKE_KEYS(Base, ...)                \
-    using mongo_odm_wrap_base = Base;                 \
-    constexpr static auto mongo_odm_mapped_fields() { \
-        return std::make_tuple(__VA_ARGS__);          \
-    }                                                 \
-    MONGO_ODM_SERIALIZE_KEYS
-
-// If using the mongo_odm::model, then also register _id as a field.
-#define MONGO_ODM_MAKE_KEYS_MODEL(Base, ...) \
-    MONGO_ODM_MAKE_KEYS(Base, MONGO_ODM_NVP(_id), __VA_ARGS__)
-
-#define MONGO_ODM_KEY(value) mongo_odm::hasCallIfFieldIsPresent<decltype(&value), &value>::call()
+#include <bsoncxx/types.hpp>
+
+#include <mongo_odm/expression_syntax.hpp>
+#include <mongo_odm/macros.hpp>
+#include <mongo_odm/util.hpp>
 
 namespace mongo_odm {
 MONGO_ODM_INLINE_NAMESPACE_BEGIN
 
+template <typename NvpT, typename T>
+class nvp_base;
+
+struct current_date_t {
+    constexpr current_date_t() {
+    }
+};
+constexpr current_date_t current_date;
+
 /**
  * An object that represents a name-value pair of a member in an object.
  * It is templated on the class of the member and its type.
  */
 template <typename Base, typename T>
-struct Nvp {
+class nvp : public nvp_base<nvp<Base, T>, T> {
+   public:
+    using type = T;
+    // In case this field is wrapped in an optional, store the underlying type.
+    using no_opt_type = remove_optional_t<T>;
+
     /**
      * Create a name-value pair from a member pointer and a name.
      * @param  t    A pointer to the member
      * @param  name The name of the member
      */
-    constexpr Nvp(T Base::*t, const char* name) : t(t), name(name) {
+    constexpr nvp(T Base::*t, const char* name) : t(t), name(name) {
+    }
+
+    /**
+     * Creates an update expression that sets the field to the given value.
+     */
+    constexpr update_expr<nvp<Base, T>, no_opt_type> operator=(const no_opt_type& val) const {
+        return {*this, val, "$set"};
+    }
+
+    /**
+    * Creates an expression that sets a date value to the current date.
+    * This is only enabled for std::chrono::time/duration values, as well as b_date.
+    */
+    template <typename U = no_opt_type>
+    constexpr typename std::enable_if_t<is_date_v<U>, current_date_expr<nvp>> operator=(
```

Comment:
You don't need typename for enable_if_t (that's the nice part of the _t).

Here and elsewhere